radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REInstruction.java

```diff
@@ -47,29 +48,31 @@
 			var buf = ByteBuffer.wrap(b.array(), pos, length);
 			return Substate.create(p, SubstateId.ofVirtualSubstate(buf));
 		}, Spin.UP, Spin.DOWN),
-		DOWN((byte) 3, (txn, i, b) -> SubstateId.fromBuffer(b), Spin.UP, Spin.DOWN),
-		LDOWN((byte) 4, (txn, i, b) -> {
-			var index = b.getInt();
-			if (index < 0 || index >= i) {
-				throw new DeserializeException("Bad local index: " + index);
-			}
-			return SubstateId.ofSubstate(txn.getId(), index);
+		VDOWNARG((byte) 3, (txn, i, b) -> {
+			int pos = b.position();
+			var p = RESerializer.deserialize(b);
+			int length = b.position() - pos;
+			var buf = ByteBuffer.wrap(b.array(), pos, length);
+			var argLength = b.get();
+			var arg = new byte[Byte.toUnsignedInt(argLength)];
+			b.get(arg);
+			return Pair.of(Substate.create(p, SubstateId.ofVirtualSubstate(buf)), arg);
 		}, Spin.UP, Spin.DOWN),
-		READ((byte) 5, (txn, i, b) -> SubstateId.fromBuffer(b), Spin.UP, Spin.UP),
-		LREAD((byte) 6, (txn, i, b) -> {
+		DOWN((byte) 4, (txn, i, b) -> SubstateId.fromBuffer(b), Spin.UP, Spin.DOWN),
+		LDOWN((byte) 5, (txn, i, b) -> {
 			var index = b.getInt();
 			if (index < 0 || index >= i) {
 				throw new DeserializeException("Bad local index: " + index);
 			}
 			return SubstateId.ofSubstate(txn.getId(), index);
-		}, Spin.UP, Spin.UP),
-		MSG((byte) 7, (txn, i, b) -> {
+		}, Spin.UP, Spin.DOWN),
+		MSG((byte) 6, (txn, i, b) -> {
 			var length = Byte.toUnsignedInt(b.get());
 			var bytes = new byte[length];
 			b.get(bytes);
 			return Hex.toHexString(bytes);
```

Comment:
I though we always used our own utils `Bytes.toHexString`, this is using BouncyCastle, is that intentional?